File: src/year2017/day6/day6.rs
//! [Day 6: Memory Reallocation](https://adventofcode.com/2017/day/6)

use rustc_hash::FxHashSet;

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (u32, u32) {
    let mut banks: Vec<u32> = data
        .split_ascii_whitespace()
        .map_while(|i| i.parse().ok())
        .collect();
    let size = u32::try_from(banks.len()).unwrap();

    let mut seen = FxHashSet::default();
    let mut iterations = 0;

    let mut part1 = 0;
    let part2;

    let mut boucle = Vec::new();

    loop {
        let state = banks.clone();

        if seen.contains(&state) {
            // we have detected a loop
            if part1 == 0 {
                part1 = iterations;
                boucle.clone_from(&state);
            } else if boucle == state {
                // count iterations within the first loop
                part2 = iterations - part1;
                break;
            }
        }

        seen.insert(state);

        // find the max
        let mut blocks_max = 0;
        let mut index_max = 0;
        for (i, &blocks) in banks.iter().enumerate() {
            if blocks_max < blocks {
                index_max = i;
                blocks_max = blocks;
            }
        }

        // redistribute blocks
        banks[index_max] = 0;
        let realloc = 1.max(blocks_max / size);
        while blocks_max > 0 {
            index_max = (index_max + 1) % banks.len();
            banks[index_max] += realloc;
            blocks_max = blocks_max.saturating_sub(realloc);
        }

        iterations += 1;
    }

    (part1, part2)
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn part1_2() {
        let puzzle = solve(TEST_INPUT);
        assert_eq!(puzzle.0, 5);
        assert_eq!(puzzle.1, 4);
    }
}

File: src/year2017/day6/day6.py
#!/usr/bin/env python3
# [Day 6: Memory Reallocation](https://adventofcode.com/2017/day/6)

import atexit
import sys
import time
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


banks = list(map(int, data.split()))
size = len(banks)

seen = set()
iterations = 0

part1 = 0
part2 = 0
loop = None

while True:
    state = ",".join(map(str, banks))

    if state in seen:
        # we have detected a loop
        if part1 == 0:
            part1 = iterations
            loop = state

        # count iterations within the first loop
        elif loop == state:
            part2 = iterations - part1
            break

    seen.add(state)

    # find the max
    blocks_max = -1
    index_max = -1
    for i, blocks in enumerate(banks):
        if blocks_max < blocks:
            index_max = i
            blocks_max = blocks

    assert blocks_max >= 1

    # redistribute blocks
    banks[index_max] = 0
    realloc = max(1, blocks_max // size)
    while blocks_max > 0:
        index_max += 1
        banks[index_max % size] += realloc
        blocks_max -= realloc

    iterations += 1


print(part1)
print(part2)
